pkg/identity/service.go

```diff
@@ -0,0 +1,187 @@
+		// authenticity, and produces a DNS-like identifier.
+		//
+		// An InvalidToken error should be returned if the provided token was not in a
+		// correct form.
+		//
+		// A NotAuthenticated error should be returned if the authenticity of the
+		// token cannot be validated.
+		Validate(context.Context, []byte) (string, error)
+	}
+
+	// InvalidToken is an error type returned by Validators to indicate that the
+	// provided authentication token was not valid.
+	InvalidToken struct{ Reason string }
+
+	// NotAuthenticated is an error type returned by Validators to indicate that the
+	// provided authentication token could not be authenticated.
+	NotAuthenticated struct{}
+)
+
+// NewService creates a new identity service.
+func NewService(v Validator, i tls.Issuer) *Service {
+	return &Service{v, i}
+}
+
+// Register registers an identity service implementation in the provided gRPC
+// server.
+func Register(g *grpc.Server, s *Service) {
+	pb.RegisterIdentityServer(g, s)
+}
+
+// Certify validates identity and signs certificates.
+func (s *Service) Certify(ctx context.Context, req *pb.CertifyRequest) (*pb.CertifyResponse, error) {
+	// Extract the relevant info from the request.
+	reqIdentity, tok, csr, err := checkRequest(req)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	if err = checkCSR(csr, reqIdentity); err != nil {
+		log.Debug()
+		return nil, status.Error(codes.FailedPrecondition, err.Error())
+	}
+
+	// Authenticate the provided token against the Kubernetes API.
+	log.Debugf("Validating token for %s", reqIdentity)
+	tokIdentity, err := s.v.Validate(ctx, tok)
+	if err != nil {
+		switch e := err.(type) {
+		case NotAuthenticated:
+			log.Infof("Authentication failed for %s: %s", reqIdentity, e)
+			return nil, status.Error(codes.FailedPrecondition, e.Error())
+		case InvalidToken:
+			log.Debugf("Invalid token provided for %s: %s", reqIdentity, e)
+			return nil, status.Error(codes.InvalidArgument, e.Error())
+		default:
+			msg := fmt.Sprintf("Error validating token for %s: %s", reqIdentity, e)
+			log.Error(msg)
+			return nil, status.Error(codes.Internal, msg)
+		}
+	}
+
+	// Ensure the requested identity matches the token's identity.
+	if reqIdentity != tokIdentity {
+		msg := fmt.Sprintf("Requested identity did not match provided token: requested=%s; found=%s",
+			reqIdentity, tokIdentity)
+		log.Debug(msg)
+		return nil, status.Error(codes.FailedPrecondition, msg)
+	}
+
+	// Create a certificate
+	crt, err := s.i.IssueEndEntityCrt(csr)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	crts := crt.ExtractRaw()
+	if len(crts) == 0 {
+		panic("We seem to have lost the certificate while signing it?")
+	}
+
+	// Bundle issuer crt with certificate so the trust path to the root can be verified.
+	log.Infof("certifying %s until %s", tokIdentity, crt.Certificate.NotAfter)
+	validUntil, err := ptypes.TimestampProto(crt.Certificate.NotAfter)
```

Comment:
There's a similar protobuf compilation issue here with this timestamp:

```
pkg/identity/service.go:127:3: cannot use validUntil (type *"github.com/linkerd/linkerd2/vendor/github.com/golang/protobuf/ptypes/timestamp".Timestamp) as type *"github.com/linkerd/linkerd2-proxy-api/vendor/github.com/golang/protobuf/ptypes/timestamp".Timestamp in field value
```

Reply:
Hmm.. How are you getting that? I can't reproduce this...
```
:; bin/lint ./pkg/...    
:; go version
go version go1.11.5 darwin/amd64
```